language-server/src/main/scala/scala/meta/languageserver/Compiler.scala

```diff
@@ -71,6 +72,23 @@ class Compiler(
         .distinct
     }
   }
+
+  def typeAt(path: AbsolutePath, line: Int, column: Int): Option[String] = {
+    val code = buffers.read(path)
+    val offset = lineColumnToOffset(code, line, column)
+    compilerByPath.get(path).map { compiler =>
+      compiler.reporter.reset()
+      val unit = compiler.newCompilationUnit(code, path.toString())
+      val richUnit = new compiler.RichCompilationUnit(unit.source)
+      compiler.unitOfFile(richUnit.source.file) = richUnit
+      val position = richUnit.position(offset)
+      val r = new Response[compiler.Tree]
+      val results = compiler.askTypeAt(position, r)
+      val typedTree = r.get.left.get
+      typedTree.tpe.toString
```

Comment:
This is the relevant bit.

In this example:

```scala
val user = User(name = "foo", age = 42)
List(user).map(x => x.name)
```

hovering on:

- `user` --> `<notype>`
- `User` --> `(name: String, age: Int)a.User`
- `List` --> `([A](xs: A*)List[A]`
- `map` --> `[B, That](f: a.User => B)(implicit bf: scala.collection.generic.CanBuildFrom[List[a.User],B,That])That`
- `x` --> `<notype>`